### PR TITLE
Implement Checkpoint / Restart for Laser Particles

### PR DESCRIPTION
**Pull request by atmyers**
Currently, laser particles are not written to / read from checkpoint files, so they do not work with restart. This implements this feature.

**Edit by @RemiLehe** : In addition, we now write the quantity `m_updated_position` to the checkpoint file. The restart test has been updated to include a laser driver.

### CODE REVIEW

File: Source/Particles/LaserParticleContainer.cpp [hunk 1]
```diff
@@ -286,6 +289,8 @@ LaserParticleContainer::InitData ()
 {
     if (!m_enabled) return;
 
+    m_laser_injection_box= Geom(0).ProbDomain();
```
Comment (atmyers) [inline Source/Particles/LaserParticleContainer.cpp]: I'm not sure that this is still needed, since we now call `Restart` instead of `InitData` when restarting.

Comment (RemiLehe) [reply Source/Particles/LaserParticleContainer.cpp]: Yes, indeed. 
However, where is the `Restart` function implemented? I am correct that this is part of `amrex` (not WarpX)?
Also: does `Restart` always call `PostRestart` under the hood?

Comment (RemiLehe) [reply Source/Particles/LaserParticleContainer.cpp]: In that case, I think that we would have to call `m_laser_injection_box= Geom(0).ProbDomain();` inside `PostRestart`.

Comment (atmyers) [reply Source/Particles/LaserParticleContainer.cpp]: `ParticleContainer::Restart` is implemented in amrex. `PostRestart` is all implemented in WarpX, and it is called by WarpX in `WarpXInitData.cpp`, after all checkpoints are read.

File: Regression/WarpX-tests.ini [hunk 2]
```diff
@@ -798,6 +798,24 @@ particleTypes = electrons
 analysisRoutine = Examples/analysis_default_regression.py
 tolerance = 1.e-14
 
+[LaserAccelerationRestart]
```
Comment (RemiLehe) [inline Regression/WarpX-tests.ini]: By default, does this test run a post-processing check, where the results of the restarted simulation are compared with the results of the original simulation?

If not, I would suggest to instead replace the beam driver of our existing restart test by a laser driver. I can help with this.

Comment (atmyers) [reply Regression/WarpX-tests.ini]: I thought it did, but now I see that, when we go through the CI, we set `doComparison=0` for all tests, which disables comparing against the original simulation. So I guess we need to do this via an analysis script, as you suggest.

Comment (RemiLehe) [reply Regression/WarpX-tests.ini]: Sounds good. I will update the PR (along with some other minor changes), hopefully within the next 2-3 hours. 
For now, I will set the PR to `[WIP]`.

File: Source/Particles/LaserParticleContainer.cpp [hunk 3]
```diff
@@ -607,6 +621,8 @@ LaserParticleContainer::PostRestart ()
...
```
Comment (RemiLehe) [inline Source/Particles/LaserParticleContainer.cpp]: Because we are not calling `InitData` anymore, I think that we do not need this line anymore.
I will remove it.

**Comment by EZoni** [comment]
I was looking into some of the restart tests, as I'm trying to improve the Python analysis script on the side, and I have a comment that might be helpful here.

The current tolerance is `sys.float_info.epsilon`, which is roughly `2e-16`. I believe if you set that to `2e-14` the test `restart_psatd`, currently failing, will pass.

On another note, I also believe that this might have something to do with the field boundary conditions. If you replace PMLs with field damping, by adding `boundary.field_lo=periodic periodic damped boundary.field_hi=periodic periodic damped` to the test's `runtime_params` in Regression/WarpX-tests.ini, I think you will not need to adjust the tolerance anymore. Field damping is indeed used in the test `restart_psatd_time_avg`, which might explain why that one isn't failing.

This might be a symptom of a separate bug related to restart combined with PMLs. I think this will become more evident once I'll have improved the Python analysis script to include checks on the field data, besides the ones we already have on the particle data. This is work in progress.

**Comment by RemiLehe** [comment]
Thanks for looking into this @EZoni. For now, I will indeed try to replace the field boundary so that we do not use the PML in this test.

**Review by RemiLehe — APPROVED** [review]
